### PR TITLE
Add support for Phabricator Plugin

### PR DESCRIPTION
**Pull request by laurentgoudet**
I'm trying to add the support for the Phabricator-Jenkins plugin, however I'm not really sure why the build fails:

```
javaposse.jobdsl.dsl.DslScriptLoaderSpec > use @Grab FAILED
    org.codehaus.groovy.control.MultipleCompilationErrorsException at DslScriptLoaderSpec.groovy:183
```

My first guess is that it's due to the plugin not being officially hosted yet, as the registration is in progress (see https://github.com/uber/phabricator-jenkins-plugin/issues/8).

Thoughts?

### CODE REVIEW

**Comment by daspilker** [comment]
I'll leave this open until the plugin is available in the update center.

File: docs/Job-reference.md [hunk 1]
```diff
@@ -4255,6 +4290,40 @@ The arguments here are in order:
...
 
+### Phabricator Notifier
+
+```groovy
+job {
+    publishers {
+        phabricatorNotifier {
+            commentOnSuccess(boolean comment = true)                // defaults to false
+            commentWithConsoleLinkOnFailure(boolean comment = true) // defaults to false
+            coberturaReportFile(String coverageFile)                // defaults to false
```
Comment (sectioneight) [inline docs/Job-reference.md]: this was removed in a recent refactor (it re-uses the results of the cobertura report, so it doesn't take any configuration)

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PhabricatorNotifierContext.groovy [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    void commentOnSuccess(boolean commentOnSuccess = true) {
+        this.commentOnSuccess = commentOnSuccess
+    }
+
+    void commentWithConsoleLinkOnFailure(boolean commentWithConsoleLinkOnFailure = true) {
+        this.commentWithConsoleLinkOnFailure = commentWithConsoleLinkOnFailure
+    }
+
+    void coberturaReportFile(String coberturaReportFile) {
```
Comment (sectioneight) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PhabricatorNotifierContext.groovy]: unused

File: docs/Job-reference.md [hunk 3]
```diff
@@ -4255,6 +4290,40 @@ The arguments here are in order:
...
+            enableUberalls(boolean enable = true)                   // defaults to false
```
Comment (sectioneight) [inline docs/Job-reference.md]: there is also a new option to drop text into a file (default `.phabricator-comment`) to post back:

https://github.com/uber/phabricator-jenkins-plugin#additional-comments

https://github.com/uber/phabricator-jenkins-plugin/blob/275cc0cf46c65305aba1e7c587967d7a8af59e20/src/main/resources/com/uber/jenkins/phabricator/PhabricatorNotifier/config.jelly#L20-L26

**Comment by sectioneight** [comment]
I just pushed 1.0 to maven... still waiting for it to show up on the wiki page https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Plugin ("No Information For This Plugin") but we're getting closer...

**Comment by laurentgoudet** [comment]
The plugin is now officially hosted (https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Plugin), and I've updated the PR based on @sectioneight feedback. Tests are still failing though, not sure if I understand how the `@RequiresPlugin` annotation is meant to work :/. Do I need to manually install the plugin with some other means in order to run the tests? Thanks!

**Comment by laurentgoudet** [comment]
So it look the `@Grab FAILED` error was due to some corrupted `grape` cache directory on my machine - I squashed & rebased, should be all good now.

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContextSpec.groovy [hunk 4]
```diff
@@ -600,6 +600,53 @@ class WrapperContextSpec extends Specification {
...
+        with(context.wrapperNodes[0]) {
+            name() == 'com.uber.jenkins.phabricator.PhabricatorBuildWrapper'
+            children().size() == 2
+            createCommit[0].value() == true
+            applyToMaster[0].value() == true
+        }
+        1 * mockJobManagement.requireMinimumPluginVersion('phabricator-plugin', '1.0')
+    }
+
+    def 'recall preBuildCleanup with minimal options'() {
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContextSpec.groovy]: why did you include this test?

Comment (laurentgoudet) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/wrapper/WrapperContextSpec.groovy]: hmm my bad, bad merge :/

**Comment by joshuaspence** [comment]
+1 on support for this plugin

**Comment by KostyaSha** [comment]
@daspilker why plugins does not implementing this on their code side?

**Comment by lanwen** [comment]
@KostyaSha Now it have [issue JENKINS-28408](https://issues.jenkins-ci.org/browse/JENKINS-28408)

**Comment by daspilker** [comment]
There have been 6 releases of the Phabricator plugin in the last 8 days. I will leave this open until the plugin has reached a stable phase. Please check whether the PR needs to updated for the latest release of the Phabricator plugin.

Or move the DSL to the Phabricator plugin, see https://github.com/jenkinsci/job-dsl-plugin/wiki/Extending-the-DSL. JENKINS-28408 will be fixed within the next couple of days.

**Comment by sectioneight** [comment]
Yea, we've been using this internally at Uber for about 6 months but it's
by no means hardened for general use yet. Hopefully as more people use it
we can iron out the kinks.

**Comment by joshuaspence** [comment]
Bump. What's the status on this? Is there any interest in having this merged?

**Comment by daspilker** [comment]
Please rebase to master, adapt to the latest version of the Phabricator plugin if necessary and adapt to the new documentation style. Otherwise I will close this pull request.

**Comment by laurentgoudet** [comment]
Sorry I got sidetracked on that. Give me few days max and I'll fix it.

**Comment by laurentgoudet** [comment]
- Rebased on master
- Updated Phabricator plugin to latest version
- Added the new plugin's options
- Updated the unit tests
- Transferred the doc to the new style

File: docs/Home.md [hunk 5]
```diff
@@ -54,6 +54,7 @@ Have a look at the [Jenkins Job DSL Gradle example](https://github.com/sheehan/j
  * Added support for the [Progress Bar Column Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Progress+Bar+Column+Plugin)
  * Added support for the [Rebuild Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Rebuild+Plugin)
  * Added support for the [Global Variable String Parameter Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Global+Variable+String+Parameter+Plugin)
+ * Added support for the [Phabricator Plugin](https://github.com/uber/phabricator-jenkins-plugin)
```
Comment (daspilker) [inline docs/Home.md]: link to the Jenkins wiki, not to GitHub

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PhabricatorNotifierContext.groovy [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+class PhabricatorNotifierContext implements Context {
+    boolean commentOnSuccess
+    boolean enableUberalls
+    String commentFile = '.phabricator-comment'
+    boolean preserveFormatting
+    int commentSize = 1000
+    boolean commentWithConsoleLinkOnFailure
+
+    void commentOnSuccess(boolean commentOnSuccess = true) {
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PhabricatorNotifierContext.groovy]: Add GroovyDoc for all DSL methods

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy [hunk 7]
```diff
@@ -1579,6 +1579,26 @@ class PublisherContext extends AbstractExtensibleContext {
     }
 
     /**
+     * Sends build status and coverage information to Pharbicator
+     *
+     * @since 1.39
+     */
+    @RequiresPlugin(id = 'phabricator-plugin', minimumVersion = '1.5')
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy]: 1.5 is not the latest version. Are you sure you want to support 1.5? The options in 1.5 do not match your code.

Comment (laurentgoudet) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy]: Hmm weird https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Plugin "Latest Release" shows 1.5 even though the last published one is 1.8.1.

**Comment by laurentgoudet** [comment]
- Fixed plugin URL
- Added GroovyDoc strings
- Fixed plugin version to v1.8.1

**Comment by daspilker** [comment]
Thanks!

@sectioneight: The commitSize option is broken, it can't be set by the DSL or by the web UI. Must be a problem in the Phabricator plugin. It's also very confusing that there are two wiki pages, https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Plugin and https://wiki.jenkins-ci.org/display/JENKINS/Phabricator+Differential+Plugin. Can you fix that?

**Comment by sectioneight** [comment]
@daspilker just got back from vacation and catching up on things.

What do you mean by "commitSize"? That's not an option provided by the plugin. Comment size? Can you tell me what the observed and expected behavior is?

As far as the wiki goes, that's been bothering me as well. It's an artifact of renaming the plugin for clarity, and triggering a bug in JIRA.

**Comment by daspilker** [comment]
@sectioneight ah, sorry, I meant commentSize. The value is always set back to 1000 after saving and reopening the config. You should send an email to the Jenkins dev list to get the wiki page deleted.

**Comment by sectioneight** [comment]
Request to delete here: https://groups.google.com/forum/#!topic/jenkinsci-dev/Nd-ErS21USA

I will look into the commentSize issue. I have never use the jobs-dsl-plugin before. Can you give me an example job configuration that reproduces this issue?

**Comment by daspilker** [comment]
You don't need the job-dsl-plugin to reproduce the issue. Just create a new job, add the "Post to Phabricator" post-build action, set "Maximum comment character length" to 2000 and click "Save". Then click "Configure" to inspect the config and notice that "Maximum comment character length" is set to 1000.

**Comment by sectioneight** [comment]
Whelp, that's embarrassing. Fix in https://github.com/uber/phabricator-jenkins-plugin/pull/105

**Pull request merged**